content/posts/vault-migration/index.md

```diff
@@ -0,0 +1,558 @@
+  ```
+
+- Restart all pods to pick up the patched configuration
+  ```bash
+  kubectl delete pod --selector=app.kubernetes.io/name=vault
+  ```
+
+### 5. Restore the snapshot in the new Vault instance
+
+With the snapshot from the old instance in hand, we restore it to the new Vault
+instance. This effectively brings the new Vault cluster to the same data state
+as the old one.
+
+- Upload the snapshot the leader pod
+
+  ```bash
+  kubectl cp "${BACKUPS_DIR}/${OLD_VAULT_SNAPSHOT_FILENAME}" \
+    "vault-${leader_index}":"${snapshot_tmp}"
+  ```
+
+- Log in with the migration token
+
+  ```bash
+  kubectl exec -it "vault-${leader_index}" -- \
+    vault login "${NEW_VAULT_MIGRATION_TOKEN}"
+  ```
+
+- Run the restore command
+  ```bash
+  kubectl exec -it "vault-${leader_index}" -- \
+    sh -c "vault operator raft snapshot restore -force \"${snapshot_tmp}\""
+  ```
+
+### 6. Configure the new Vault cluster to use the old unseal key
+
+To allow the restored Vault data to function properly, we need to configure the
+new instance to use the same unseal key as the old Vault. This step modifies the
+config map and restarts the pods to apply the change.
+
+- Patch the configuration of the new Vault cluster with the credentials of the
+  old Azure Key Vault
+
+  ```bash
+  kubectl get cm vault-config -o yaml > "${vault_config_filename}"
+
+  ed '+/seal "/' "${vault_config_filename}" <<EOF
+  .,/}/d
+  -
+  a
+      seal "azurekeyvault" {
+        tenant_id      = "${TENANT_ID}"
+        client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
+        vault_name     = "${OLD_KEY_VAULT_NAME}"
+        key_name       = "${OLD_VAULT_UNSEAL_KEY_SECRET_NAME}"
+      }
+  .
+  w
+  q
+  EOF
+
+  kubectl apply -f "${vault_config_filename}"
+  ```
+
+- Restart all pods to pick up the patched configuration
+  ```bash
+  kubectl delete pod --selector=app.kubernetes.io/name=vault
+  ```
+
+### 7. Prepare the unseal key migration on the new Vault instance
+
+To enable unseal key migration, we first need to configure the new Vault cluster
+to support dual seal configuration — keeping the old seal block active but
+disabled, and adding a new one that points to the new Azure Key Vault.
+
+- Update the Vault configuration:
+
+```bash
+kubectl get cm vault-config -o yaml > vault-config.yaml
+
+ed '+/seal "/' vault-config.yaml <<EOF
+a
+      disabled       = true
+.
+/}/
+a
+    seal "azurekeyvault" {
+      tenant_id      = "${TENANT_ID}"
+      client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
+      vault_name     = "${NEW_KEY_VAULT_NAME}"
+      key_name       = "${NEW_VAULT_UNSEAL_KEY_SECRET_NAME}"
+    }
+.
+w
+q
+EOF
+
+kubectl apply -f vault-config.yaml
+```
+
+### 8. Perform the unseal key migration on the follower pods
+
+Vault only allows unseal key migration on follower nodes, so we restart each
+follower pod and run the migration steps.
+
+- Iterate through each follower pod and migrate using all three unseal keys:
+
+  ```bash
+  for i in {0..2}; do
```

Comment:
Replace with pod selection by labels

Reply:
Can I leave that to you, @theseion? I no longer have a setup to test this effectively. Alternatively, we can just leave it as is for the blog post. It works and it is also quite easy to understand.

Reply:
I'll take care of it. Goes hand in hand with the query to the StatefulSet at the top.